src/main/java/org/wildfly/security/ssl/SSLUtils.java

```diff
@@ -103,23 +106,29 @@ private SSLUtils() {}
                 for (Provider.Service service : services) {
                     if (serviceType.equals(service.getType())) {
                         String protocolName = service.getAlgorithm();
-                        if (!preferredProviderByAlgorithm.containsKey(protocolName)) {
-                            preferredProviderByAlgorithm.put(protocolName, provider);
-                        }
+                        List<Provider> providerList = preferredProviderByAlgorithm.computeIfAbsent(protocolName, s -> new ArrayList<>());
+                        providerList.add(provider);
                     }
                 }
             }
         }
 
-        // now figure out the supported protocol set.
-
-        String[] supportedProtocols = protocolSelector.evaluate(preferredProviderByAlgorithm.keySet().toArray(NO_STRINGS));
-        for (String supportedProtocol : supportedProtocols) {
-            Provider provider = preferredProviderByAlgorithm.get(supportedProtocol);
-            if (provider != null) {
-                return createSimpleSslContextFactory(supportedProtocol, provider);
-            }
+        // now return a factory that will return the best match is can create.
+        final String[] supportedProtocols = protocolSelector.evaluate(preferredProviderByAlgorithm.keySet().toArray(NO_STRINGS));
+        if (supportedProtocols.length > 0) {
+            return () -> {
+                for (String protocol : supportedProtocols) {
+                    List<Provider> providerList = preferredProviderByAlgorithm.getOrDefault(protocol, Collections.emptyList());
```

Comment:
Minor thing, but maybe {{preferredProviderByAlgorithm}} should be renamed to {{preferredProviderByProtocol}}